drivers/sensor/bmp388/bmp388.c

```diff
@@ -0,0 +1,789 @@
+/* Bosch BMP388 pressure sensor
+ *
+ * Copyright (c) 2020 Facebook, Inc. and its affiliates
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Datasheet:
+ * https://www.bosch-sensortec.com/media/boschsensortec/downloads/datasheets/bst-bmp388-ds001.pdf
+ */
+
+#define DT_DRV_COMPAT bosch_bmp388
+
+#include <logging/log.h>
+#include <sys/byteorder.h>
+#include <drivers/i2c.h>
+#include <drivers/sensor.h>
+
+#include "bmp388.h"
+
+LOG_MODULE_REGISTER(BMP388, CONFIG_SENSOR_LOG_LEVEL);
+
+#if defined(CONFIG_BMP388_ODR_RUNTIME)
+static const struct {
+	uint16_t freq_int;
+	uint16_t freq_milli;
+} bmp388_odr_map[] = {
+	{ 0, 3 },       /* 25/8192 - 327.68s */
+	{ 0, 6 },       /* 25/4096 - 163.84s */
+	{ 0, 12 },      /* 25/2048 - 81.92s */
+	{ 0, 24 },      /* 25/1024 - 40.96s */
+	{ 0, 49 },      /* 25/512 - 20.48s */
+	{ 0, 98 },      /* 25/256 - 10.24s */
+	{ 0, 195 },     /* 25/128 - 5.12s */
+	{ 0, 391 },     /* 25/64 - 2.56s */
+	{ 0, 781 },     /* 25/32 - 1.28s */
+	{ 1, 563 },     /* 25/16 - 640ms */
+	{ 3, 125 },     /* 25/8 - 320ms */
+	{ 6, 250 },     /* 25/4 - 160ms */
+	{ 12, 500 },    /* 25/2 - 80ms */
+	{ 25, 0 },      /* 25 - 40ms */
+	{ 50, 0 },      /* 50 - 20ms */
+	{ 100, 0 },     /* 100 - 10ms */
+	{ 200, 0 },     /* 200 - 5ms */
+};
+#endif
+
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(spi)
+static int bmp388_transceive(const struct device *dev,
+			     void *data, size_t length)
+{
+	const struct bmp388_config *cfg = DEV_CFG(dev);
+	const struct spi_buf buf = { .buf = data, .len = length };
+	const struct spi_buf_set s = { .buffers = &buf, .count = 1 };
+
+	return spi_transceive(cfg->bus, &cfg->spi_cfg, &s, &s);
+}
+
+static int bmp388_read_spi(const struct device *dev,
+			   uint8_t reg,
+			   void *data,
+			   size_t length)
+{
+	const struct bmp388_config *cfg = DEV_CFG(dev);
+
+	/* Reads must clock out a dummy byte after sending the address. */
+	uint8_t reg_buf[2] = { reg | BIT(7), 0 };
+	const struct spi_buf buf[2] = {
+		{ .buf = reg_buf, .len = 2 },
+		{ .buf = data, .len = length }
+	};
+	const struct spi_buf_set tx = { .buffers = buf, .count = 1 };
+	const struct spi_buf_set rx = { .buffers = buf, .count = 2 };
```

Comment:
Is this case correct? Not too many bytes skipped?

Reply:
Yes, see comment above about dummy byte.

Reply:
So to read from SPI you perform a TX transaction (2 bytes) with reg + 1 DUMMY, and then again same TX transaction followed by RX transaction of length byte. Am I missing something about spi_transceive()?

Reply:
To me this is the last point not clear (but if it works maybe I'm missing something). All the rest of open points seem good.

Reply:
TX/RX happen at the same time, so we write/read two bytes (we don't care about read bytes but use the same buffer for efficiency), then we read N bytes. This works in my STM32L4 based setup.